astropy/io/fits/hdu/compressed/compressed.py

```diff
@@ -572,408 +526,146 @@ def data(self):
-                _shape.reverse()
-                _shape = tuple(_shape)
-                _format = _format[_format.rfind(".") + 1 :]
-
-        # if data is not touched yet, use header info.
-        else:
-            _shape = ()
-
-            for idx in range(self.header["NAXIS"]):
-                _shape += (self.header["NAXIS" + str(idx + 1)],)
+        if self.data is None:
+            return
 
-            _format = BITPIX2DTYPE[self.header["BITPIX"]]
+        heap = compress_image_data(
+            self.data, self.compression_type, bintable.header, bintable.columns
+        )
 
-        return (self.name, self.ver, class_name, len(self.header), _shape, _format)
+        dtype = bintable.columns.dtype.newbyteorder(">")
+        buf = np.frombuffer(heap, dtype=np.uint8)
+        data = (
+            buf[: bintable._theap]
+            .view(dtype=dtype, type=np.rec.recarray)
+            .view(FITS_rec)
+        )
+        data._load_variable_length_data = False
+        data._coldefs = bintable.columns
+        data._heapoffset = bintable._theap
+        data._heapsize = len(buf) - bintable._theap
 
-    def _update_compressed_data(self):
-        """
-        Compress the image data so that it may be written to a file.
-        """
-        # Check to see that the image_header matches the image data
-        image_bitpix = DTYPE2BITPIX[self.data.dtype.name]
-
-        if image_bitpix != self._orig_bitpix or self.data.shape != self.shape:
-            self._update_header_data(self.header)
-
-        # TODO: This is copied right out of _ImageBaseHDU._writedata_internal;
-        # it would be cool if we could use an internal ImageHDU and use that to
-        # write to a buffer for compression or something. See ticket #88
-        # deal with unsigned integer 16, 32 and 64 data
-        old_data = self.data
-        if _is_pseudo_integer(self.data.dtype):
-            # Convert the unsigned array to signed
-            self.data = np.array(
-                self.data - _pseudo_zero(self.data.dtype),
-                dtype=f"=i{self.data.dtype.itemsize}",
-            )
+        bintable.data = data
 
-        try:
-            nrows = self._header["NAXIS2"]
-            tbsize = self._header["NAXIS1"] * nrows
+    @property
+    def _hdu_modified_from_disk(self):
+        return self._bintable is None or self.header._modified or self._data_modified
 
-            self._header["PCOUNT"] = 0
-            if "THEAP" in self._header:
-                del self._header["THEAP"]
-            self._theap = tbsize
+    def _prewriteto(self, checksum=False, inplace=False):
+        # In some cases the user might do hdu.data.shape = (...) to change the
+        # shape of the data. In this case, we reset the data on the image HDU
+        # to force the header to be updated and to set _data_modified since
+        # the data will need to be recompressed from scratch.
+        if (
+            not self._data_modified
```

Comment:
Yes you are absolutely right, these lines are not actually needed - the reason it works fine without is because when writing out a file that has been written, even in update mode, ``ImageHDU.update_header`` already checks to make sure the data shape hasn't changed. We even have a test for this behavior (``test_open_scaled_in_update_mode_compressed``) so have removed the lines.

Reply:
> Do we support inplace updates for the VLA columns used for compression?

I don't think we ever supported this, nor do we need to. I don't think anyone has ever asked for the ability to recompress only part of the image, so I think what we have for now is fine.

Reply:
Just a note that I called the private attribute ``_data_modifed`` as opposed to just ``_modified`` as done in say ``ImageHDU`` because I wanted to make it clear it only concerned the data and not the header. For the header we do rely on ``.modified``.

Reply:
> >  Do we support inplace updates for the VLA columns used for compression?
>
> I don't think we ever supported this, nor do we need to. I don't think anyone has ever asked for the ability to recompress only part of the image, so I think what we have for now is fine.

That's what I was thinking but got confused by some changes/comments. It's related to the other discussion, about when the full array is decompressed and must recompressed.

Reply:
This should work, it does on main (but seems not tested ?):

```In [1]: %astropy
Numpy 2.1.1
Astropy 7.0.0.dev321+gcacde01abc

In [2]: hdul = fits.open("comp.fits", mode="update")

In [3]: data = hdul[1].data

In [4]: data
Out[4]: 
array([[  7,   7,   7, ...,  91,  86,  78],
       [  7,   7,   7, ...,  94,  90,  84],
       [  7,   6,   6, ...,  98,  94,  89],
       ...,
       [  1,   0,   0, ..., 107,  97,  81],
       [  1,   0,   1, ..., 100,  90,  72],
       [  1,   1,   1, ...,  84,  75,  65]], dtype=int16)

In [5]: data[0] = 0

In [6]: hdul[1]._hdu_modified_from_disk
Out[6]: False

In [7]: hdul.flush()

In [8]: hdul.close()

In [9]: fits.getdata("comp.fits")
Out[9]: 
array([[  7,   7,   7, ...,  91,  86,  78],
       [  7,   7,   7, ...,  94,  90,  84],
       [  7,   6,   6, ...,  98,  94,  89],
       ...,
       [  1,   0,   0, ..., 107,  97,  81],
       [  1,   0,   1, ..., 100,  90,  72],
       [  1,   1,   1, ...,  84,  75,  65]], dtype=int16)
```